### ISSUE TITLE
De-duplicate storage of proposals 

### ISSUE THREAD

**Issue by jumaffre** [microsoft/CCF#2466]
When a member submits a proposal, the proposal itself is stored twice: in the `public:ccf.gov.js.proposals` table and in the governance history table (`public:ccf.gov.history`) as the body of the signed request. This isn't ideal, especially for large proposals. Instead, the `request_body` field of the `SignedReq` could be changed to a field which indicates which table and key to read to get the corresponding request body.

**Comment by eddyashton** [comment]
I think there's a risking of making it harder to independently verify signed governance commands. Currently I think you should be able to look at a raw request in `gov.history` and verify the signature over it, entirely independent of any interpretation/knowledge of other tables? And this would introduce a tight binding that you're able to use some data from `gov.history` and some additional bytes from another table and combine them into the (verbatim) original request that was signed? I think if we're going to de-duplicate like this we should try to have an entire field in another place in the KV, not part of it (ie - storing a HTTP request in `js.proposals`, with associated re-parsing, rather than just the body).

This duplication is also mostly a concern for the ledger, not the in-memory KV. We expect that shortly after a member submits a proposal, they submit a vote for it, and that vote overwrites their entry in `gov.history`. I think we could improve this further and make `gov.history` a mono-keyed/singleton/what-are-we-calling-it-today table, storing only a single value at `key 0` rather than one-per-member?

**Comment by achamayou** [comment]
It's actually worse than duplicating, it's really triplicating in bad cases (JS apps).

**Comment by jumaffre** [comment]
@eddyashton, I agree that auditing the governance history would require looking up another key in the raw proposal table (or in the proposals info for a vote, if we want to do that too), but I can't find a strong reason why the coupling is a bad thing at this point. Parsing the ledger is now easy enough and substituting the raw body of the request with the value of a key in another table seems straightforward enough. 

Good point about the in-memory vs. ledger concerns. My other assumption was that it would limit the size of the ledger entry sent over the ring buffer and the size of the snapshot at that particular point. But maybe this isn't as big of an issue once we get https://github.com/microsoft/CCF/issues/1858? 

@achamayou, in the case of the JS apps, this seems trickier as the app itself isn't quite the whole proposal (only its `args`).

**Comment by eddyashton** [comment]
@jumaffre This will also introduce coupling in the code, where "record this request in the voting history" is no longer a separable function but needs to write some complex state dependent on the rest of the proposal's execution. In particular I guess it needs to happen _after_ the proposal's been written, to know what the proposal ID is? So it'd have to be represented as something like:

```
table "public:ccf.gov.history" (1 write):
      33c7ddf1fd6741161e3cba5e23387b6bd824b285a2ddd9027262f461660a7c20:
        {
          "key_id": "33c7ddf1fd6741161e3cba5e23387b6bd824b285a2ddd9027262f461660a7c20",
          "md": "SHA256",
          "req": "KHJlcXVlc3QtdGFyZ2V0KTogcG9zdCAvZ292L3Byb3Bvc2Fscy5qcwpkaWdlc3Q6IFNIQS0yNTY9amF2UlpyYzErQlJXaUZITXZlWmhYRGxjTmU1VGs3MGJEWTg2NXVWMkcrZz0KY29udGVudC1sZW5ndGg6IDU5NDQ=",

          // This is the one that has changed, and should be read as "concatenate these bits together"
          "signed_request": [
              {
                "raw": "HJlcXVlc3QtdGFyZ2V0KTogcG9zdC..."
              },
              {
                "ref": {
                  "map_name": "public:ccf.gov.proposals",
                  "key": "deadbeef" // This isn't known at the point we currently try to write to gov.history
                 }
              }
          ],

          "sig": "MGYCMQDszdWwgABi+IfZjRQBpdmf2L3auIRCvXZnHPG+KXBff3mYhHRrDdVEjL2WJWHSjHYCMQCoUTiQW6ep9R+zUao0oVUGtot85mbI04A19"
        }


```

**Comment by jumaffre** [comment]
Unless I'm missing places where the governance history is recorded, I believe the proposal ID is always known by that point, e.g.:
https://github.com/microsoft/CCF/blob/c4a81abff847651f782ebe675912b7bc01fa4191/src/node/rpc/member_frontend.h#L1040-L1045

**Comment by eddyashton** [comment]
Ha, you're right! I was thinking signed-request storage was still a very early general stage in `frontend.h`, but of course its purely managed by `member_frontend` now so it can (and does!) happen after `proposal_id` is known. Ignore my last comment.
